database/manager.go

```diff
@@ -0,0 +1,267 @@
+// See the file LICENSE for licensing terms.
+
+package database
+
+import (
+	"container/heap"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"go.uber.org/zap"
+)
+
+//
+// DatabaseManager writes all committed keys to the database on a timer.
+//
+
+type DatabaseManager struct {
+	logger      logging.Logger
+	keyManagers map[RelayerID]*keyManager
+	interval    time.Duration
+	db          RelayerDatabase
+}
+
+func NewDatabaseManager(logger logging.Logger, db RelayerDatabase, interval time.Duration, keys []RelayerID) *DatabaseManager {
+	keyManagers := make(map[RelayerID]*keyManager)
+	for _, key := range keys {
+		keyManager := newKeyManager(logger, key)
+		go keyManager.run()
+		keyManagers[key] = keyManager
+	}
+	return &DatabaseManager{
+		logger:      logger,
+		db:          db,
+		interval:    interval,
+		keyManagers: keyManagers,
+	}
+}
+
+// Run writes all committed keys to the database on the configured timer.
+// This function should only be called once.
+func (dm *DatabaseManager) Run() {
+	for range time.Tick(dm.interval) {
+		for id, km := range dm.keyManagers {
+			// Ensure we're not writing the default value
+			km.lock.RLock()
+			currMaxCommittedHeight := km.maxCommittedHeight
+			km.lock.RUnlock()
+			if currMaxCommittedHeight == 0 {
+				continue
+			}
+			storedHeight, err := getLatestProcessedBlockHeight(dm.db, id)
+			if err != nil && !IsKeyNotFoundError(err) {
+				dm.logger.Error(
+					"Failed to get latest processed block height",
+					zap.Error(err),
+					zap.String("relayerID", id.ID.String()),
+				)
+				continue
+			}
+			if storedHeight >= currMaxCommittedHeight {
+				continue
+			}
+			dm.logger.Debug(
+				"Writing height",
+				zap.Uint64("height", currMaxCommittedHeight),
+				zap.String("relayerID", id.ID.String()),
+			)
+			err = dm.db.Put(id.ID, LatestProcessedBlockKey, []byte(strconv.FormatUint(currMaxCommittedHeight, 10)))
+			if err != nil {
+				dm.logger.Error(
+					"Failed to write latest processed block height",
+					zap.Error(err),
+					zap.String("relayerID", id.ID.String()),
+				)
+				continue
+			}
+		}
+	}
+}
+
+func (dm *DatabaseManager) Get(id RelayerID, key DataKey) ([]byte, error) {
+	return dm.db.Get(id.ID, key)
+}
+
+func (dm *DatabaseManager) PrepareHeight(id RelayerID, height uint64, totalMessages uint64) error {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		dm.logger.Error("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return fmt.Errorf("key manager not found")
+	}
+	km.prepareHeight(height, totalMessages)
+	return nil
+}
+
+func (dm *DatabaseManager) Finished(id RelayerID, height uint64) {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		dm.logger.Error("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return
+	}
+	km.finished <- height
+}
+
+// intHeap adapted from https://pkg.go.dev/container/heap#example-package-IntHeap
+type intHeap []uint64
```

Comment:
Can we rename this to `uint64Heap`? The typecheck on line 119 will panic if we provide any other type of integer.

Reply:
Alternatively, there's a generic heap package [here](https://pkg.go.dev/github.com/fufuok/utils/generic/heap#section-readme)